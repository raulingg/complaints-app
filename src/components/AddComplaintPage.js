import React, { Component } from 'react';
import { connect } from 'react-redux';
import ComplaintFormComponent from './ComplaintForm';
import { startAddComplaint as startAddComplaintAction } from '../actions/complaints';

export class AddComplaintPage extends Component {
  onSubmit = complaint => {
    const { history, startAddComplaint } = this.props;
    startAddComplaint(complaint);
    history.push('/');
  };

  render() {
    return (
      <div>
        <h1>Reporta tu denuncia </h1>
        <ComplaintFormComponent onSubmit={this.onSubmit} />
      </div>
    );
  }
}
const mapStateToProps = (state, props) => ({
  complaint: state.complaints.find(complaint => complaint.id === props.match.params.id),
});

const mapDispatchToProps = dispatch => ({
  startAddComplaint: complaint => dispatch(startAddComplaintAction(complaint)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AddComplaintPage);
