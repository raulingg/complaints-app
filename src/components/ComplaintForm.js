import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { startAddComplaint as startAddComplaintAction } from '../actions/complaints';

export class ComplaintForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.complaint ? props.complaint.title : '',
      content: props.complaint ? props.complaint.content : '',
      reportTo: props.complaint ? props.complaint.reportTo : '',
      happenedAt: props.complaint ? moment(props.complaint.happenedAt) : moment(),
    };
  }

  onTitleChange = e => {
    const title = e.target.value;
    this.setState(() => ({ title }));
  };

  onContentChange = e => {
    const content = e.target.value;
    this.setState(() => ({ content }));
  };

  onReportToChange = e => {
    const reportTo = e.target.value;
    this.setState(() => ({ reportTo }));
  };

  onHappenedAtChange = e => {
    const happenedAt = e.target.value;
    this.setState(() => ({ happenedAt }));
  };

  onSubmit = e => {
    e.preventDefault();
    const { title, content, happenedAt, reportTo } = this.state;
    const { startAddComplaint, history } = this.props;
    startAddComplaint({
      title,
      content,
      happenedAt,
      reportTo,
    });
    history.push('/');
  };

  render() {
    const { title, content, happenedAt, reportTo } = this.state;
    return (
      <form onSubmit={this.onSubmit}>
        <div>
          <input type="text" name="title" value={title} onChange={this.onTitleChange} />
        </div>
        <div>
          <textarea
            name="content"
            maxLength="2000"
            rows="10"
            value={content}
            onChange={this.onContentChange}
          />
        </div>
        <div>
          <input
            type="date"
            name="happenedAt"
            value={happenedAt}
            onChange={this.onHappenedAtChange}
          />
        </div>
        <div>
          <input type="text" name="reportTo" value={reportTo} onChange={this.onReportToChange} />
        </div>
        <div>
          <button className="button">Guardar cambios</button>
        </div>
      </form>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddComplaint: complaint => dispatch(startAddComplaintAction(complaint)),
});

export default connect(
  undefined,
  mapDispatchToProps
)(ComplaintForm);
