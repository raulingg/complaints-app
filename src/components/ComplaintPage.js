import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

export const ComplaintPage = ({ complaint }) => (
  <div>
    <h1>{complaint.title}</h1>
    <p>
      Denuncia: <span>{complaint.reportTo}</span>
    </p>
    <div dangerouslySetInnerHTML={{ __html: complaint.content }} />
    <p>
      <span>Denunciado el {moment(complaint.happenedAt).format()}</span>
    </p>
  </div>
);

const mapStateToProps = (state, props) => ({
  complaint: state.complaints.find(complaint => String(complaint.id) === props.match.params.id),
});

export default connect(mapStateToProps)(ComplaintPage);
