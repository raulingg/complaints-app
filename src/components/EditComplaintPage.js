import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { compose } from 'recompose';
import ComplaintFormComponent from './ComplaintForm';
import { startEditComplaint as startEditComplaintAction } from '../actions/complaints';
import withAuthorization from './auth/withAuthorization';

export class EditComplaintPage extends Component {
  onSubmit = updates => {
    const {
      history,
      startEditComplaint,
      complaint: { id },
    } = this.props;
    startEditComplaint(id, updates);
    history.push('/');
  };

  render() {
    const { complaint } = this.props;
    return (
      <div>
        <h1>Editar denuncia </h1>
        <ComplaintFormComponent complaint={complaint} onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  complaint: state.complaints.find(complaint => String(complaint.id) === props.match.params.id),
});

const mapDispatchToProps = dispatch => ({
  startEditComplaint: (id, complaint) => dispatch(startEditComplaintAction(id, complaint)),
});

export default compose(
  withRouter,
  withAuthorization,
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(EditComplaintPage);
