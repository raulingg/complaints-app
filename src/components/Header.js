import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout as startLogoutAction, startLogin as startLoginAction } from '../actions/auth';

export const Header = ({ isAuthenticated, startLogout, startLogin }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/">
          <h1>Denuncias</h1>
        </Link>
        {isAuthenticated ? (
          <button className="button button--link" onClick={startLogout}>
            Salir
          </button>
        ) : (
          <button className="button button--link" onClick={startLogin}>
            Iniciar sesi√≥n
          </button>
        )}
      </div>
    </div>
  </header>
);

const mapStateToProps = state => ({
  isAuthenticated: !!state.auth.uid,
});

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogoutAction()),
  startLogin: () => dispatch(startLoginAction()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Header);
