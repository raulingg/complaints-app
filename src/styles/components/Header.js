import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { colors, sizes } from '../settings';

const Header = styled.header`
  background: ${colors.darkBlue};
`;

const Content = styled.div`
  align-items: center;
  display: flex;
  justify-content: space-between;
  padding: ${sizes.s} 0;
`;

const Title = styled(Link)`
  color: white;
  text-decoration: none;
  h1 {
    margin: 0;
  }
`;

export { Header, Content, Title };
