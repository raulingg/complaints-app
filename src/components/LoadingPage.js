import React from 'react';
import * as LoaderStyles from '../styles/components/Loader';

const LoadingPage = () => (
  <LoaderStyles.Loader>
    <LoaderStyles.Image src="images/loader.gif" alt="load spinner" />
  </LoaderStyles.Loader>
);

export default LoadingPage;
