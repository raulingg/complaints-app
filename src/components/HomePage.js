import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Button from '../styles/components/Buttons';
import { startSetComplaints as startSetComplaintsAction } from '../actions/complaints';
import LoadingPage from './LoadingPage';

export class HomePage extends Component {
  state = {
    isLoading: true,
  };

  async componentDidMount() {
    const { startSetComplaints } = this.props;
    await startSetComplaints();
    this.setState({ isLoading: false });
  }

  render() {
    const { complaints } = this.props;
    const { isLoading } = this.state;

    return isLoading ? (
      <LoadingPage />
    ) : (
      <div>
        <div>
          <div>Denuncia</div>
          <div>Tipo</div>
          <div>Fecha</div>
        </div>
        <div>
          <Button as={Link} to="/add">
            Denunciar
          </Button>
          {complaints.length === 0 ? (
            <div>
              <span>No encontramos ninguna denuncia.</span>
            </div>
          ) : (
            complaints.map(complaint => (
              <div key={complaint.id}>
                <p>
                  <Link to={`show/${complaint.id}`}>{complaint.title}</Link>
                </p>
              </div>
            ))
          )}
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  complaints: state.complaints,
});

const mapDispatchToProps = dispatch => ({
  startSetComplaints: () => dispatch(startSetComplaintsAction()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(HomePage);
