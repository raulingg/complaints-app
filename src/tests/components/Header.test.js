import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';

test('should render Header correctly', () => {
  const wrapper = shallow(<Header startLogout={() => {}} startLogin={() => {}} isAuthenticated />);
  expect(wrapper).toMatchSnapshot();
});

test('should call startLogout on button click when user is authenticated', () => {
  const startLogout = jest.fn();
  const isAuthenticated = true;

  const wrapper = shallow(<Header startLogout={startLogout} isAuthenticated={isAuthenticated} />);
  wrapper.find('button').simulate('click');
  expect(startLogout).toHaveBeenCalled();
});

test('should call startLogin on button click when user is unauthenticated', () => {
  const startLogin = jest.fn();
  const isAuthenticated = false;

  const wrapper = shallow(<Header startLogin={startLogin} isAuthenticated={isAuthenticated} />);
  wrapper.find('button').simulate('click');
  expect(startLogin).toHaveBeenCalled();
});
