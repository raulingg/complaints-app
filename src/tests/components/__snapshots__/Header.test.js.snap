// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header
  className="header"
>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        replace={false}
        to="/"
      >
        <h1>
          Denuncias
        </h1>
      </Link>
      <button
        className="button button--link"
        onClick={[Function]}
      >
        Salir
      </button>
    </div>
  </div>
</header>
`;
